openassessment/xblock/code_executor/executors/mixins.py

```diff
@@ -0,0 +1,223 @@
+    COMPILE_COMMAND_TEMPLATE = '{source_file} {executable_file}'
+    RUN_COMMAND_STDIN_INPUT_TEMPLATE = '{executable_file}'
+    RUN_COMMAND_FILE_INPUT_TEMPLATE = '{executable_file} {input_file}'
+
+    @classmethod
+    def get_config(cls) -> dict:
+        return {
+            'id': cls.id,
+            'display_name': cls.display_name,
+            'language': cls.language,
+            'version': cls.version,
+            'profiles': [
+                epicbox.Profile(
+                    name='{}-compile'.format(cls.id), docker_image=cls.docker_image,
+                ),
+                epicbox.Profile(
+                    name='{}-run'.format(cls.id),
+                    docker_image=cls.docker_image,
+                    user='sandbox',
+                    read_only=True,
+                ),
+            ],
+        }
+
+    def __init__(
+        self,
+        source_code: str,
+        files: List[Dict[str, bytes]] = [],
+        limits: Dict[str, int] = None,
+        **kwargs
+    ) -> None:
+        self.compile_profile_name = '{}-compile'.format(self.id)
+        self.run_profile_name = '{}-run'.format(self.id)
+
+        self.source_file_name = self.SOURCE_FILE_NAME_TEMPLATE.format(name=uuid4())
+        self.executable_file_name = self.EXECUTABLE_FILE_NAME_TEMPLATE.format(name=uuid4())
+        self.limits = limits
+        self.files = files + [
+            {'name': self.source_file_name, 'content': bytes(source_code, 'utf-8'),}
+        ]
+        self.working_directory_context_manager = epicbox.working_directory()
+        self.working_directory = None
+
+    def __enter__(self):
+        self.working_directory = self.working_directory_context_manager.__enter__()
+
+        # Compile code.
+        results = epicbox.run(
+            self.compile_profile_name,
+            command=self.COMPILE_COMMAND_TEMPLATE.format(
+                source_file=self.source_file_name, executable_file=self.executable_file_name,
+            ),
+            files=self.files,
+            workdir=self.working_directory,
+            limits=self.limits,
+        )
+        if results['exit_code'] != 0:
+            message = results['stderr'].decode('utf-8')
+            if results['timeout']:
+                message = 'Compilation time limit exceeded.'
+            elif results['oom_killed']:
+                message = 'Compilation memory limit exceeded.'
+
+            raise CodeCompilationError(message=message)
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        self.working_directory_context_manager.__exit__(exc_type, exc_value, exc_tb)
+
+    def _check_working_directory_creation(self):
+        if self.working_directory is None:
+            raise WorkingDirectoryNotInitializedException(
+                'Working directory not created. '
+                'Please use code executor with a context manager.'
+            )
+
+    def run_input(self, input: str) -> dict:
+        self._check_working_directory_creation()
```

Comment:
To optimize performance, code is compiled only once. The executable is placed inside the working directory. We take the executable from inside that working directory when we run the input.

For scripted languages, I have a different reason. The code_executor package does not know how it will be used. It has been intentionally made independent. As such, the `run_input` here does not know if any files were placed in the working directory. 

A working directory is always created, regardless of whether or not any files are passed to the constructor. This proves beneficial for forward compatibility, storing utility files (like data), and reducing complexity. This will also allow us to use multi-file code in the future.